### PR TITLE
Add closing event

### PR DESCRIPTION
**Pull request by Nargonath**
Closes (#4194)

### CODE REVIEW

File: API.md [hunk 1]
```diff
@@ -595,6 +595,18 @@ server.events.on('start', () => {
 });
 ```
 
+##### <a name="server.events.closing" /> `'closing'` Event
+
+The `'closing'` event type is emitted when the server is stopped using [`server.stop()`](#server.stop()) but before all underlying connections have been closed thus before the [`'stop'`](#server.events.stop) event is triggered.
```
Comment (cjihrig) [inline API.md:600]:
```suggestion
The `'closing'` event type is emitted when the server is stopped using [`server.stop()`](#server.stop()) but before all underlying connections have been closed, and thus before the [`'stop'`](#server.events.stop) event is triggered.
```

**Review by cjihrig — APPROVED** [review]

File: API.md (hunk 1 above)
Comment (Nargonath) [reply API.md]: Thanks for your suggestion.

**Review by kanongil — COMMENTED** [review]
Looks good.

I would probably explain that it triggers when incoming connections are no longer accepted, but active request are still being processed.

A test to verify this would also be nice, since there could be a timing issue between the call to `listener.close()` and it actually applying to the underlying socket. Especially, if we later change the `emit()` method to immediately trigger, as suggested in #4184.

**Comment by devinivy** [comment]
Looks good!  If you're going to make any changes to the tests, I might also recommend supporting this statement from the docs with a test:
> and thus before the [`'stop'`](#server.events.stop) event is triggered

In my opinion it could be a part of the same test you already have written.

**Comment by Nargonath** [comment]
Thanks @kanongil for your feedbacks. i'll add a test for that.

Yeah good idea Devin. I'll update my test at the same time.

**Comment by Nargonath** [comment]
@kanongil I have added another test. Just to make sure I understand properly what you suggested, you wanted a test to make sure the events emission order was consistent even though there is an active request while the server is stopped. Also making sure that a denied incoming request would not trigger the closing event. Am I right?

**Comment by Nargonath** [comment]
@kanongil have you got any time to take a look at the second test I added? Is it what you were looking for?

**Comment by Nargonath** [comment]
Is there anything blocking this PR? Should we proceed and merge it? If you think it needs some rework, please feel free to voice your concerns. 😃

**Review by devinivy — APPROVED** [review]

**Comment by Nargonath** [comment]
Alright I'll merge this. The feature works and if we want to improve the tests suite later on we can still come back to it. Thanks everyone for your feedbacks.

**Pull request merged**